homeassistant/components/tplink/entity.py

```diff
@@ -62,24 +121,294 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+            "Device feature: %s (%s) needs an entity description defined in HA",
+            feature.name,
+            feature.id,
         )
+        return None
+
+    @classmethod
+    def _entities_for_device[
+        _E: CoordinatedTPLinkFeatureEntity,
+        _D: TPLinkFeatureEntityDescription,
+    ](
+        cls,
+        device: Device,
+        coordinator: TPLinkDataUpdateCoordinator,
+        *,
+        feature_type: Feature.Type,
+        entity_class: type[_E],
+        descriptions: Mapping[str, _D],
+        parent: Device | None = None,
+    ) -> list[_E]:
+        """Return a list of entities to add.
+
+        This filters out unwanted features to avoid creating unnecessary entities
+        for device features that are implemented by specialized platforms like light.
+        """
+        entities: list[_E] = [
+            entity_class(
+                device,
+                coordinator,
+                feature=feat,
+                description=desc,
+                parent=parent,
+            )
+            for feat in device.features.values()
+            if feat.type == feature_type
+            and feat.id not in EXCLUDED_FEATURES
+            and (
+                feat.category is not Feature.Category.Primary
+                or device.device_type not in DEVICETYPES_WITH_SPECIALIZED_PLATFORMS
+            )
+            and (
+                desc := cls._description_for_feature(
+                    feat, descriptions, child_alias=device.alias if parent else None
+                )
+            )
+        ]
+        return entities
+
+    @classmethod
+    def entities_for_device_and_its_children[
+        _E: CoordinatedTPLinkFeatureEntity,
+        _D: TPLinkFeatureEntityDescription,
+    ](
+        cls,
+        device: Device,
+        coordinator: TPLinkDataUpdateCoordinator,
+        *,
+        feature_type: Feature.Type,
+        entity_class: type[_E],
+        descriptions: Mapping[str, _D],
+        child_coordinators: list[TPLinkDataUpdateCoordinator] | None = None,
+    ) -> list[_E]:
+        """Create entities for device and its children.
+
+        This is a helper that calls *_entities_for_device* for the device and its children.
+        """
+        entities: list[_E] = []
+        # Add parent entities before children so via_device id works.
+        entities.extend(
+            cls._entities_for_device(
+                device,
+                coordinator=coordinator,
+                feature_type=feature_type,
+                entity_class=entity_class,
+                descriptions=descriptions,
+            )
+        )
+        if device.children:
+            _LOGGER.debug("Initializing device with %s children", len(device.children))
+            for idx, child in enumerate(device.children):
+                # HS300 does not like too many concurrent requests and its emeter data requires
+                # a request for each socket, so we receive separate coordinators.
```

Comment:
Please split the long comment

Reply:
Done in https://github.com/home-assistant/core/pull/120404